### DISCUSSION TITLE
Improving the hygiene of 1Password CLI 2 integration account specification

### DISCUSSION THREAD

**Discussion by halostatue** [twpayne/chezmoi#1974] [Ideas]
The recent release of 1Password CLI 2.0.0 changes a number of things, but the two biggest changes are:

1. `op signin` requires the `--account` param. This has been fixed by #1961 and #1966, although not yet released.
2. When  using `op signin` without biometric authentication, the environment variable used is no longer `OP_SESSION_<accountShorthand>`, but `OP_SESSION_<accountUUID>`. That means that if I have an entry like `{{ opDocument foo '' myaccount }}`, the session cannot be inherited from the environment because chezmoi will try `OP_SESSION_myaccount`, but the session token will actually be stored in `OP_SESSION_myaccountUUID`. If `1password.prompt` is `false`, then the UUID is the only value that can be safely used in `op*` template functions.

A solution would be to cache the account maps at the beginning of a session (on the first call to any of the `op*` template functions). This would be retrieved with `op account list --format=json`, which returns responses like this:

```json
[
  { "url": "account1.1password.com", "email": "user@account1.example.com", "user_uuid": "account1UUID" },
  { "url": "account2.1password.ca", "email": "user@account2.example.com", "user_uuid": "account2UUID" }
]
```

I _believe_ that this should be transformed into a lookup table with the following rules:

- the full URL is mapped to the UUID;
- the first word of the URL (`account1.1password.com` becomes `account1`) is mapped to the UUID;
- the email address is mapped to the UUID;
- the UUID is mapped to the UUID; and
- entries that collide for any reason are removed (see below).

The example given above would be transformed into the following (as JSON):

```json
{
  "account1": "account1UUID",
  "account1.1password.com": "account1UUID",
  "user@account1.example.com": "account1UUID",
  "account1UUID": "account1UUID",
  "account2": "account2UUID",
  "account2.1password.ca": "account2UUID",
  "user@account2.example.com": "account2UUID",
  "account2UUID": "account2UUID"
}
```

I _believe_ that any of these values could be used to specify the `--account` parameter, but it would probably require testing to be sure.

Collisions should be removed because they will result in ambiguous results. I believe that the order of `op accounts list --format=json` is _stable_, but is the reverse order of sign-in (that is, the first account is _last_ in the list). Known collision cases would be:

1. `myaccount.1password.com` and `myaccount.1password.ca` or `myaccount.1password.eu` exist and the person has accounts on both (meaning that the short version of the URL would collide), OR
2. `user@example.com` exists in both `account1` and `account2`.

In theory, I guess it’s possible for the full URL to collide as well, but in that case, the email address would definitely not collide.

This doesn’t really affect me at the moment as I’m using biometric authentication, which is _all kinds of awesome_ for chezmoi, but I think that it will be a problem for a fairly large class of users of 1Password CLI 2.0. I’m not sure how quickly this could be implemented, but I think that some sort of solution (as I’m not getting any feedback from [Agilebits themselves](https://1password.community/discussion/comment/634067/#Comment_634067)) is necessary.

**Comment by twpayne** [comment]
Thanks very much for opening this discussion. Personally, I'm only using 1Password with a single (work) account so my knowledge of the details is much less than yours.

This all sounds good. Note that on my machine running `op account list --format=json` also includes a `shorthand` field that can also be mapped to an account UUID.

**Comment by halostatue** [comment]
What is the output of your `op --version`? Mine is `2.0.0`.

**Comment by twpayne** [comment]
I also have `op --version` equal to `2.0.0`. I'm on macOS.

**Comment by halostatue** [comment]
Well, ain’t that just peachy. So far, the 1Password folks seem to be ignoring my thread on 1password.community, but it appears that there is a difference in output for `op account list --format=json` depending on whether biometric authentication is enabled. IMO, this is a bug in op.

I am using biometric authentication (14" MBP with M1 Max), so the keys for the account list are `url`, `email`, and `user_uuid`. You are not, so the keys for the account list are `url`, `email`, `user_uuid`, and `shorthand`.

The logic that I described is still sound with one tweak:

-  `url` is mapped to `user_uuid`;
- the first word before the dots of `url` (`account1.1password.com` becomes `account1`) is mapped to `user_uuid`;
-  `email` is mapped to `user_uuid`;
- `uuid` is mapped to `user_uuid`;
- entries that collide are removed as described; and
- `shorthand` is mapped to `user_uuid`, if present.

Shorthand overrides the collision detection mode because it is the well-known shorthand that should be used. We could also _potentially_ add the first word before the `@` of `email` is mapped to `user_uuid`.

What do you think?

**Reply by halostatue** [reply]
@twpayne I’m going to start on a draft PR that can look this stuff up. The plan is to do the following:

- `url -> user_uuid`
- `firstword(url) -> user_uuid`
- `email -> user_uuid`
- `firstword(email) -> user_uuid`
- `firstword(email)@url -> user_uuid`
- `user_uuid ->  user_uuid`
- remove collisions
- `shorthand -> user_uuid` (if present)

My first version of this is likely to require serious review, and I’m not sure what the best way to test this would be, so I’ll be submitting it as a draft. Additionally, my Golang is fairly…I don’t use the language often. But I think we should actually implement this and move forward.

**Comment by antonalekseev** [comment]
I'm not sure it belongs here, but have noticed another caveat. My `op --version` is `2.0.2`, but that doesn't mean that `op item create ...` will necessarily produce an item of `"version": 2`, in my case it's still `"version": 1`. Therefore `chezmoi execute-template "{{ onepasswordItemFields \"$UUID\" | toJson }}" | jq .` returns `{}`. Probably it happens due to the fact that I still have `1Password 7` macOS GUI app which is incompatible with `"version": 2`.

**Reply by halostatue** [reply]
This feels like something separate, and may be worth opening as a discussion with <1password.community> as well as opening a bug here, because it _sounds_ like even though you’re using `op 2.x`, you’re getting output that would nominally be compatible with `op 1.x`. Is that correct?

**Reply by SimonBarendse** [reply]
Hi all, 1Password developer here. :wave: The `version` field in output of `op item` commands indicates the version of the item itself (i.e. the number of times the item was edited). It is not an indicator for the format of the output. The format of the output will be determined by your `op --version` and will remain backwards compatible in all 2.y.z versions. I hope this helps. Let me know if there's anything else I can clarify or if there's any changes you'd like in the behavior of the 1Password CLI.

**Reply by SimonBarendse** [reply]
FWIW: we're working on documentation for the output JSON of all CLI commands so that there's a place where you can find and refer to this information. I have highlighted this friction point you encountered with the folks working on this documentation effort.

**Reply by halostatue** [reply]
@antonalekseev, what @SimonBarendse has said here suggests that there’s either a bug in `op`’s output or in `chezmoi`’s handling of the existing output when connected to 1P7. I have fully switched to 1P8, so that’s not something that I’m going to be easily able to debug.

**Reply by twpayne** [reply]
@antonalekseev Please can you post the output with chezmoi's `--debug` flag. This will show exactly how chezmoi is invoking `op`. You can then run those commands without chezmoi to see how `op` is behaving. Something like:

```console
$ chezmoi execute-template --debug "{{ onepasswordItemFields \"$UUID\" | toJson }}" | jq
```

**Reply by antonalekseev** [reply]
@SimonBarendse Ah, I see. Sorry for drawing misleading conclusions RE `version` attribute. Of course, documentation for the JSON output would be greatly appreciated.

@twpayne Sure. Here is an item created for testing purposes

```
$ op item get --format json testitem | jq -rM '.id + " " + .category + " " + .fields[0].value'
oyzzjv6kpnf2xf3oincqkk72ci LOGIN testuser
```

```
$ chezmoi execute-template "{{ onepasswordItemFields \"oyzzjv6kpnf2xf3oincqkk72ci\" | toJson }}" | jq .
{}
```

Retrying the above with `--debug` flag yields tons of output, but the most interesting part in my opinion is

```
2022-04-27T08:27:08+03:00 INF Output args=["op","--version"] duration=25.302816ms output="2.0.2\n" path=/usr/local/bin/op size=6
2022-04-27T08:27:08+03:00 INF IdempotentCmdOutput args=["op","--version"] component=system output="2.0.2\n" path=/usr/local/bin/op
2022-04-27T08:27:09+03:00 INF Output args=["op","--session","hKujmOm6xL8Cl8NxxoguM2S474ZzEVmvkfGyh1d4BKo","item","get","--format","json","oyzzjv6kpnf2xf3oincqkk72ci"] duration=1.233045936s output="{\n  \"id\": \"oyzzjv6kpnf2xf3oincqkk72ci\",\n  \"title\": \"testitem\",\n ..." path=/usr/local/bin/op size=941
2022-04-27T08:27:09+03:00 INF IdempotentCmdOutput args=["op","--session","hKujmOm6xL8Cl8NxxoguM2S474ZzEVmvkfGyh1d4BKo","item","get","--format","json","oyzzjv6kpnf2xf3oincqkk72ci"] component=system output="{\n  \"id\": \"oyzzjv6kpnf2xf3oincqkk72ci\",\n  \"title\": \"testitem\",\n ..." path=/usr/local/bin/op
2022-04-27T08:27:09+03:00 INF Close component=persistentState
```
@halostatue I'm still on macOS `10.14.6` which was EOSed a year ago. 1P7 is still supported AFAIK, but not sure that on `10.14.6`. It would be probably better for me to upgrade, than to spend developer's time on debugging obscure issues on an obsolete OS version.

```
$ uname -a
Darwin pat.local 18.7.0 Darwin Kernel Version 18.7.0: Tue Jun 22 19:37:08 PDT 2021; root:xnu-4903.278.70~1/RELEASE_X86_64 x86_64
```

**Reply by halostatue** [reply]
Can you compare the output of

```console
$ chezmoi execute-template "{{ onepassword \"$UUID\") | toJson }}" | jq .
```

I believe that `onepasswordItemFields` is no longer as useful with the op v2 data format. The transforms made still make sense, but fewer items seem to produce usable values IMO. It _may_ need a deeper restructuring than has been done to date (and _might_ be better as a _filter_ function than an execution function).

**Reply by antonalekseev** [reply]
Yes this one yields the proper output. The only thing is that data structure used for `fields` is not convenient for use in templates and requires either writing boilerplate code like suggested here https://github.com/twpayne/chezmoi/issues/474#issuecomment-586157328 or hard coding `fields` indices. `jsonpath`-like function would help, but unfortunately is not supported by `chezmoi` as far as I know.

**Comment by SimonBarendse** [comment]
Thank you @halostatue for providing such a detailed overview of the problem! I see quite a few friction points in this thread, so going to look and see if there's any changes we can make on the 1Password side to make this easier on your end. I'll get back to you all here.

Cheers, Simon (Developer, Developer Products @ 1Password)

**Reply by SimonBarendse** [reply]
Hi folks, just dove in the code. Could you confirm for me whether the code at https://github.com/twpayne/chezmoi/blob/58f9f451bc089e31b0f22ea51cef338123536b13/pkg/cmd/onepasswordtemplatefuncs.go#L221 reading the `OP_SESSION_x` variable is to determine whether or not there's a need to run `op signin` to authenticate the account if it isn't already?

I believe an ideal scenario would be to remove your need to manage the 1Password credential via chezmoi code and have the 1Password CLI determine whether authentication is necessary and prompt the user in this case. Do you share this goal?

When using the new Biometric Unlock feature that authenticates the CLI using 1Password 8, I believe this is already the case. 1Password CLI will detect whether or not it is already authenticated, and prompt the user for authentication if still needed.

We have the desire to achieve the same for accounts authenticated directly on the terminal. When running any command (e.g. `op item get`) and you're not yet authenticated, instead of erroring and having to run `op signin` separately, that command would prompt for your account password.
To keep the session token available for subsequent commands invoked from the same terminal, we're currently using the parent environment to store the session token (the `OP_SESSION_x` environment variables). This is no longer an option if we'd like any command to have the ability to prompt, as it would require wrapping all your commands in `eval`. Therefore, there's some investigating that we'd still need to carry out to see if there's feasible alternatives to pass the session token between invocations of 1Password CLI commands that match our security needs. I'm thinking for example about using a daemon process and securing it to limit access to a single terminal process.

Wondering if the above is the underlying issue you're running into and if it would help resolve some of the friction points you're running into.

I acknowledge this doesn't resolve the friction you experienced around configuring the account to use (using an argument in v1 and `--account` in v2), shall we discuss that separately?

**Reply by SimonBarendse** [reply]
> This doesn’t really affect me at the moment as I’m using biometric authentication, which is all kinds of awesome for chezmoi, but I think that it will be a problem for a fairly large class of users of 1Password CLI 2.0

For what it's worth: the overwhelming majority of feedback we're receiving is along these lines. We have been receiving a ton of positive feedback on the Biometric Unlock feature and almost all users we're talking with are using it. I'm curious to learn if you have heard of different experiences and if so, if you have insight in reasons and/or friction points that make Biometric Unlock not work in these cases. We believe the integration with the unlock integration with the 1Password client app offers the best experience between the two options and I'd love to remove any hurdles if that means we can get more folks to experience that better experience.

**Reply by SimonBarendse** [reply]
And to add some additional transparency on our roadmap: so far we haven't heard of a need for authentication directly in the terminal nor about users using that anymore in v2. For this reason we haven't yet started/prioritized the above suggested improvement to how/when users are prompted for their password and have instead focused on areas where users have reported needs for improvement. If you do see a need for this, we can re-evaluate its priority accordingly of course (without making any promises, there's a ton of exciting stuff we've planned for Developer Products and I don't want to set any expectations I cannot uphold).

**Reply by halostatue** [reply]
> Hi folks, just dove in the code. Could you confirm for me whether the code … reading the `OP_SESSION_x` variable is to determine whether or not there's a need to run `op signin` to authenticate the account if it isn't already?

That is correct.

> I believe an ideal scenario would be to remove your need to manage the 1Password credential via chezmoi code and have the 1Password CLI determine whether authentication is necessary and prompt the user in this case. Do you share this goal?

Mmm. There is a relatively recently introduced configuration option that **prevents** chezmoi from prompting performing `op signin` if `OP_SESSION_x` is not present. There are some users who would want to authenticate once in the terminal and then run `chezmoi apply`. Since I could be running chezmoi on a remote server (where biometric authentication isn’t possible), I would far prefer that case than where I get a seemingly random prompt while `chezmoi` is running.

> When using the new Biometric Unlock feature that authenticates the CLI using 1Password 8, I believe this is already the case. 1Password CLI will detect whether or not it is already authenticated, and prompt the user for authentication if still needed.

That is correct. The biometric unlock is _spectacular_, but not everyone has a T2-enabled or Apple Silicon Mac where this would be viable.

> We have the desire to achieve the same for accounts authenticated directly on the terminal. When running any command (e.g. `op item get`) and you're not yet authenticated, instead of erroring and having to run `op signin` separately, that command would prompt for your account password. To keep the session token available for subsequent commands invoked from the same terminal, we're currently using the parent environment to store the session token (the `OP_SESSION_x` environment variables). This is no longer an option if we'd like any command to have the ability to prompt, as it would require wrapping all your commands in `eval`. Therefore, there's some investigating that we'd still need to carry out to see if there's feasible alternatives to pass the session token between invocations of 1Password CLI commands that match our security needs. I'm thinking for example about using a daemon process and securing it to limit access to a single terminal process.

From a security perspective, I’m not sure that this would be the right way to handle that for command-line input. I can trust the biometric request dialog on macOS, but I don’t know that I’d want to consider trusting a command-line prompt, since any application can put any text to a terminal and there’s no way to verify it as authentic. I would prefer 1Password fail a request if not authenticated, unless I explicitly ask for auto-login (e.g., `op item get foo --auto-login`).

> I acknowledge this doesn't resolve the friction you experienced around configuring the account to use (using an argument in v1 and `--account` in v2), shall we discuss that separately?

This also doesn’t really resolve the friction of _identifying_ the account to be used. I’ll reply separately in this discussion because I think that it’s worth pulling an example that I used in the 1P community discussion.

**Reply by SimonBarendse** [reply]
> Since I could be running chezmoi on a remote server (where biometric authentication isn’t possible)

This is still early stage, but we're exploring the introduction of separate "service account" credentials that you could pass with e.g. `OP_CREDENTIAL` that can authenticate services and other remote processes. This effort currently has two main goals:
1. Offer an authentication mechanism that requires no input from a human
2. Allow separately managing access for the service, i.e. it should not require access to everything you as a developer have access to. So you can apply the principle of least privilege in these places.

>  could be running chezmoi on a remote server (where biometric authentication isn’t possible), I would far prefer that case than where I get a seemingly random prompt while chezmoi is running

1Password CLI has checks build in to determine if it can prompt (a.o. if tty is available), if it cannot prompt it will return an error.

> not everyone has a T2-enabled or Apple Silicon Mac

Other than the name suggests, Biometric Unlock should also work when you do not have biometrics available. Instead of prompting for your biometrics, it will request your password. Does this help in your case?

> I can trust the biometric request dialog on macOS, but I don’t know that I’d want to consider trusting a command-line prompt, since any application can put any text to a terminal and there’s no way to verify it as authentic.

This is a great insight! For anything that builds on top of the CLI, there should be a way for users to validate it's the 1Password CLI prompting for their password. I believe the above of 1Password apps also prompting in the case biometrics aren't available is a step in that direction.

With a password prompt via 1Password 8. If we'd add `OP_CREDENTIAL` for headless jobs, do you see any gaps left?

**Reply by halostatue** [reply]
Mmm. `OP_CREDENTIAL` sounds interesting, but note the use case described (`chezmoi apply` on a server reached via SSH) is not headless *as such*. Being able to do the following in such a scenario:

```console
$ ssh remotehost
remotehost $ eval $(op signin --account myaccount)
Password: •••••••••
remotehost $ chezmoi apply
…
…
$
```

Is *far* better than:

```console
$ ssh remotehost
remotehost $ chezmoi apply
…
Password:
```

I don’t know what I’m entering the password *for* in the second scenario. If `eval $(op signin)` exports `OP_CREDENTIAL` the same way that it currently exports `OP_SESSION_<uuid>` (and it used to export `OP_SESSION_<shorthand>`), then we’re still at a usable point. But we can’t necessarily expect that `op` will be able to tell a GUI to show a password prompt. (That’s also a great question: can / should `op` be used without the GUI application installed? I know that there’s a push to use secrets automation, but this isn’t *quite* the same sort of scenario where secrets automation would be useful—and I believe that chezmoi would need development work to support secrets automation as opposed to `op` integration).

**Reply by SimonBarendse** [reply]
> Mmm. OP_CREDENTIAL sounds interesting, but note the use case described (chezmoi apply on a server reached via SSH) is not headless as such.

Ah gotcha. Chezmoi on a remote server is operated by a real person. In those cases, logging in as a user continues to be the way to go. And `eval $(op signin)` as you propose is the way to do so on machines where the 1Password application is not installed.

---

Let me see if I can summarize what we need to achieve and make a proposal. Please let me know if there's additional goals you'd add.

## Goals

#### For Chezmoi & 1Password users
1. Chezmoi users can use any of the [supported account identifiers](https://developer.1password.com/docs/cli/use-multiple-accounts/) (Sign in Address, account ID, user ID) to indicate which 1Password account they want to use.
1. Chezmoi users can use 1Password authentication method of their preference (directly on Terminal, using the 1Password application, and any other options we may add in the future)

#### For Chezmoi maintainers & contributors
I'd like to propose the following goals to make the Chezmoi integration with 1Password easier to maintain.

1. **Chezmoi is agnositc of how 1Password stores authentication details**. Two reasons for this: 1) it's a pain to manage in Chezmoi and adds a lot of complexity and a big reason for this discussion to happen in the first place. 2) We have not built the storage of authentication details to be depended on by users or integrations. We consider the storage an _internal implementation detail_ to achieve the goals of the external `op signin` and `op signout` interface. Not every `op signin` will result in an `OP_SESSION_x` envvar to be set. For example when the integration with the 1Password app is enabled, the session is not stored in an environment variable. Session storage may change at any time without notice, as it is not considered part of the v2 interface. Building Chezmoi so that it depends on just the v2 interface ensures Chezmoi continues to be compatible with all future 1Password CLI v2 releases. This ensures a better experience for users (no old versions we have to deprecate early and without notice) and a better experience for maintainers (no code changes required).
1. **It is clearer which parts of 1Password CLI are built to depend on**. We (1Password) should make it clearer which parts of 1Password CLI we consider part of the public interface you can depend on, and which parts are internal implementation details that may change at any time without notice.
1. **Chezmoi is agnostic of the method of authentication used** for 1Password by users (e.g. directly on terminal, using 1Password app using biometrics, using 1Password app using password etc.). We may add additional methods of authentication to 1Password to provide a better experience for users, similar to how we've done so with biometrics with the 1Password app. When we do add new methods of authentication and Chezmoi is agnostic of the method of authentication used, these new methods are immediately supported for Chezmoi + 1Password users without any changes to the Chezmoi code needed. For example, we could consider in the future (we don't have any plans for this currently) to look into how we can improve the authentication experience for running jobs remotely as a user. We could look into using the locally running 1password app to authenticate 1Password CLI running on a remote. When Chezmoi is agnostic of the method of auth used, this will work with Chezmoi out-of-the-box with no changes needed! 🎉

## Proposal

### Chezmoi is agnositc of how 1Password stores authentication details

I see two possible solutions here:
1. Chezmoi always runs `op signin` independent of whether the user is already signed in or not. `op signin` is idempotent, so it will only prompt the user for authentication if they're not already authenticated. I'll make sure we add this to the documentation.
2. Chezmoi uses `op whoami` to check if the user is already authenticated. As noted above this shouldn't be needed to check, but you can rely on `op whoami` to exit without error when the user is authenticated and with error when they're not.

### It is clearer which parts of 1Password CLI are built to depend on

I believe we need to do better here. We're considering `OP_SESSION_x` envvar an internal implementation detail that shouldn't be depended on directly. I'd love to hear suggestions if you have any of how we can better communicate what is and isn't part of the interface you can rely on being stable. We're always looking to learn and improve.

We consider everything documented on https://developer.1password.com part of the v2 contract that we keep backwards compatible. Everything that is not documented, please let us know so that we can either add it to the docs and make sure we keep it backwards compatible within v2 or we can built an alternative way to achieve what you need that does not depend on internals.

I hope the above goal of making Chezmoi agnostic of `OP_SESSION_x` clarifies that we do not consider this part of the interface we keep backwards compatible. I'll make a proposal below for an alternative.

### Chezmoi is agnostic of the method of authentication used

I believe we get this for free once we remove the depedence on `OP_SESSION_x` with one of the above two approaches (idempotent `op signin` or `op whoami`). New methods of authentication added to 1Password will immediately be supported by Chezmoi 🎉

---

@halostatue 
1. Do you have any additions on the goals?
2. How do you feel about the proposal? Would that work?

**Reply by twpayne** [reply]
Thank you very much @SimonBarendse for thinking this through in so much detail.

From a chezmoi maintainers point of view, this makes a lot of sense: the more that chezmoi is insulated from 1Password CLI internals, the less burden there is on both projects.

@halostatue, as an expert chezmoi and 1Password user, what are your thoughts?

**Reply by halostatue** [reply]
I have not had time to look at it yet. I’ll keep this notification unread and get back in a day or two.

**Reply by gunzy83** [reply]
Thanks for the great write up @SimonBarendse, I think goals section makes perfect sense and as a user point `1.` actually matches how I am trying to use it (long time chezmoi user, newer to 1password).

After some experimenting and getting over the issue I mentioned in #2353, I think the idempotent `op signin` or `op whoami` (ideally scoped to the account specified in the template, particularly if you have multiple accounts) would ensure if we are logged in already we can run chezmoi without prompting or interruption. If the user wants to pre-signin, that is up to them otherwise they get a standard `op signin` prompt on the first occurence.

I am interested in what @halostatue has to say as his use-cases seem a bit more advanced than mine.

**Reply by halostatue** [reply]
Hi, @SimonBarendse, and sorry for the long delay in response.

Overall, the plan sounds fine. The lack of a customizable account shorthand when using biometrics means that we will absolutely need to implement the account name aliasing described elsewhere in this discussion, because using either the User ID or Account ID values is opaque and unreadable, and the sign-in address (whether the email or the URL) *could* be non-unique (I could have two unique user accounts on the same 1P URL, or use the same email address on two different 1P URLs). We can *also* consider setting up an explicit Chezmoi configuration option to allow for manual mapping.

I do have one concern about the *always run `op signin`* option that I *cannot* test without disabling biometric login, and that’s whether running `op signin` in a subshell would carry the authorization tracking upwards. Essentially, Chezmoi is doing the following for `op signin`:

```sh
# …process stuff and find a case where we need `op` interactions
op signin --account ${accountId}
op document get …

# …more processing. Another `op` interaction:
op signin --account ${accountId}
op document get …
```

In the biometric case, this is *not* an issue. In the non-biometric case, at least with the versions of `op` and 1Password 7 or 8, the output of `op signin` is to be `eval`ed because it sets `OP_SESSION_x`.

It *might* be easier for us to rely on `op signin --raw --account X`. If we get a non-zero-length value back, we can store the token and pass it with `--session TOKEN` on all calls to `op`. Otherwise, we will need to act sort of like `direnv` does.

As long as `op signin --raw` and `op … --session TOKEN` will continue to work, then I don’t think that there’s any issue. Otherwise, the fact that `op signin` only works by setting environment variables means that Chezmoi needs to be at least partially aware of them or it cannot call `op signin` during execution.

**Reply by halostatue** [reply]
I *believe* that this discussion is mostly resolved with the following two points:

1. In #2425 and #2440, we have implemented the 1Password account map. `pkg/cmd/onepasswordtemplatefuncs_test.go:TestOnepasswordAccountMap(t *testing.T)` is have the best description of the fine details of the behaviour. It it likely that we will be able to take that and put into the 1Password template functions documentation.
2. We are *already* using `op signin --raw --account accountUUID` and `op … --session TOKEN`. If `op signin --raw --account accountUUID` is made to work in an idempotent manner when `OP_SESSION_x` is set, then everything works as expected.

While `op whoami` would be a nice thing to have, if successive calls to `op signin` are idempotent (generating something new—and therefore prompting—*only* if the session in the environment has expired) is sufficient to our purposes.

@SimonBarendse, I think we have things working *very* nicely now.

**Reply by SimonBarendse** [reply]
> As long as op signin --raw and op … --session TOKEN will continue to work

Yes, you can rely on these remaining backwards compatible for all `v2.x.y` versions. :+1: We won't remove any flags.

**Comment by halostatue** [comment]
@SimonBarendse, the way that `op` gets used in chezmoi is through template functions. I mostly use the `onepasswordDocument` template function (corresponding to `op document get …`), but there are other functions available. You can test these with chezmoi as follows:

```console
$ chezmoi execute-template "{{ onepasswordItemFields \"$UUID\" | toJson }}"
$ chezmoi execute-template "{{ onepasswordDocument \"$UUID\" }}"
```

I use document fragments to construct my `~/.aws/credentials` file and loop over a set of them. Essentially, I end up filling a template with values like this: `{{ onepasswordDocument "aws-credentials: personal" "Personal" "myaccount" }}`. (The arguments are `document name/id`, `vault name/id`, `account name/id`.) This turns into an `op` command looking like `op document get 'aws-credentials: personal' --vault Personal --account myaccount`.

As long as op sees that my account has a computed (or set) shorthand of `myaccount`, everything is fine. If, however, with `op` v1, I had set the shorthand to `myaccount` for account `foobarbaz.1password.ca`, then all of a sudden my chezmoi templates no longer work under biometric authentication (because the `shorthand` value is no longer used or respected). I have to either switch my account parameters to `foobarbaz` or the account UUID.

If I’m not using biometric authentication, then `OP_SESSION_myaccount` is no longer set, so chezmoi has to prompt me for login (if I have that enabled; if I do _not_ have that enabled, I can no longer login unless I explicitly do `export OP_SESSION_myaccount=$OP_SESSION_{uuid}`). Then, because the shorthand still exists, my account reference of `myaccount` still exists.

There’s at least two issues here:
1. Biometric authentication no longer exposes a user-configurable shorthand for an account. If my shorthand matched my domain prefix, there’s nothing visible. But if my domain prefix was `really-awesome-company.1password.eu`,but I had a shorthand of `rac`…I now have to change my configurations.
2. Because the `OP_SESSION_` variable name has changed, non-biometric users can’t _disable_ auto-login and have chezmoi Just Work as they could previously. The first post in this discussion (and a follow-up later) suggests a way that chezmoi can work around this to discover the correct `OP_SESSION_` variable through heuristics (it would also let us use the discovered value for `op --account`).

It’s complex and messy and I’m probably not doing a great job of describing it, but these small changes by 1Password CLI have rippled into 3–4 different areas that require workarounds or fixes from 1Password.

I’m not saying that all of the changes made are bad (I do think that requiring `--account` on `op signin` is a wrong choice, as I described in the 1password.community thread), but they have impacts and some of them are subtle.

**Reply by SimonBarendse** [reply]
Is there a reason you don't want to use:
```
{{ onepasswordDocument "aws-credentials: personal" "Personal" "foobarbaz.1password.ca" }}
```

The main reason we dropped the concept of a "shorthand" in 1Password version 2 is that we wanted to remove confusion that occurs from the shorthand potentially deviating from what the account is known as otherwise. There's already quite a few concepts for to the sign-in (sign-in address, secret key, password, email-address) and we wanted to remove one to reduce the cognitive overhead required to get started with, use and debug the sign-in process for users. The fact that these shorthands do not exist in 1Password 8 adds to the complexity there'd be in keeping them around (as there's no longer an intuitive place to look them up).

I hope this background provides some insight to our reasoning, I'm looking forward to learn your perspective on this!

> I do think that requiring --account on op signin is a wrong choice, as I described in the 1password.community thread

Thank you for providing that reasoning on the forum thread, I'm really appreciating all your feedback and thorough descriptions. I've replied on the forum to this.

**Reply by halostatue** [reply]
I can think of two reasons:

1. Until you just mentioned it, it was not clear to me that you could use the full sign-in domain name as the account name (which means we would need to add that to the logic above, @twpayne);
2. Backwards compatibility. As I said, I had to update my templates because of the upgrade to op v2, and the reasoning for such was unclear.

Ultimately, I think that a lot of what’s going on is that the changes have been made, but the reasoning behind those changes has not been communicated so that downstream projects *like chezmoi* could communicate that to their users, even with just a link. This might be because 1Password wasn’t aware all of the places / ways that people were using `op`.

**Reply by jkyjustin** [reply]
Apologies for changing the internals of the session keys, there were some technical implications of no longer being able to use the account shorthand as the env var key with biometric auth, and I share your sentiment that we should have communicated that better.

**Comment by SimonBarendse** [comment]
Hi folks, I'm sorry for my slow replies here. I'm really strapped for time and I want this thread to receive the attention it deserves, so I have asked a teammate, @jkyjustin, to jump in here to help out. You're in great hands!

**Comment by jkyjustin** [comment]
Hey @halostatue 

> In theory, I guess it’s possible for the full URL to collide as well, but in that case, the email address would definitely not collide.

Regarding this, yes the full URL can collide. One example of this is a 1Password employee testing across multiple CLI users on a single device - so when specifying the `--account` flag on `signin` commands, using the `account_uuid` or `shorthand` fields are preferred.

I believe of the two, `account_uuid` is a safer choice, as `shorthand` will not be printed as a response of `op accout list --format json` when biometrics are enabled, so whereas `account_uuid` is always printed.

**Reply by halostatue** [reply]
I still think that you should enable sort of shorthand for the accounts even with biometrics enabled; if I am signed in as `austin@mycompany.com` *and* `admin@mycompany.com` (and have different accesses), then I’d like to be able to simply say `--account admin` *even* if I’m using biometrics. I know it’s a bit harder, but I think that it could be managed by an extra field (eventually exposed on the non-CLI UIs) that is set by default but can be changed.

**Reply by jkyjustin** [reply]
Hey @halostatue 

I absolutely agree that that would be the ideal behavior, however, because when auth states are shared across the CLI and the desktop applications with biometric authentication enabled, shorthands currently have no meaning to the desktop clients.

While it would be possible to achieve this by getting the desktop clients to implement some sort of meaning to a shorthand (it was previously only a concept for the CLI), we won't be able to use it for biometric auth to switch between accounts of the same signin URL.

**Reply by twpayne** [reply]
Replying to bump the conversation. I want chezmoi to have first class 1Password support.

**Comment by halostatue** [comment]
Opened draft #2425 for the 1Password account lookup behaviour.
